pkg/storage/gcs/gcp_labels_tags.go

```diff
@@ -170,120 +187,195 @@ func filterTagList(ctx context.Context, client *rscmgr.TagBindingsClient, parent
+	parent := fmt.Sprintf(bucketParentPathFmt, cr.Spec.Storage.GCS.Bucket)
+	tags, err := t.getTagsToBind(ctx, parent)
+	if err != nil {
+		return err
+	}
+	if len(tags) <= 0 {
+		return NoTagsDefined
+	}
+
+	if err := t.tagBindingsClient.CreateTagBindings(ctx, parent, tags); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getTagClientOptions returns the tag client options adding the credentials and
+// the endpoint which will be used by the client.
+func getTagClientOptions(listers *regopclient.StorageListers, endpoint string) ([]option.ClientOption, error) {
+	cfg, err := GetConfig(listers)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read GCS configuration for creating tag client: %w", err)
+	}
+
+	opts := []option.ClientOption{
+		option.WithCredentialsJSON([]byte(cfg.KeyfileData)),
+		option.WithEndpoint(endpoint),
+	}
+
+	return opts, nil
+}
+
+// getTagBindingsClient returns the client to be used for creating tag bindings to
+// the resources.
+func getTagBindingsClient(ctx context.Context, listers *regopclient.StorageListers, location string) (TagBindingsService, error) {
+	endpoint := fmt.Sprintf("https://%s-%s", location, resourceManagerHostSubPath)
+	opts, err := getTagClientOptions(listers, endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create tag binding client options: %w", err)
+	}
+
+	client, err := rscmgr.NewTagBindingsRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create tag binding client: %w", err)
+	}
+	return &tagBindingsClient{client}, nil
+}
+
+// NewTagManager creates a tagServiceManager instance. Explicit Close() must
+// be called when tag service is no longer needed to Close() all clients.
+func NewTagManager(ctx context.Context, listers *regopclient.StorageListers, region string) (TagService, error) {
+	client, err := getTagBindingsClient(ctx, listers, region)
+	if err != nil || client == nil {
+		return nil, err
+	}
+	mgr := &tagServiceManager{
+		Listers:           listers,
+		tagBindingsClient: client,
+	}
+	return mgr, nil
+}
+
+func (c *tagBindingsClient) Close() {
+	if err := c.TagBindingsClient.Close(); err != nil {
+		klog.Errorf("failed to close tag binding client: %v", err)
+	}
+}
+
+func (t *tagServiceManager) Close() {
+	t.tagBindingsClient.Close()
+}
+
+// AddTagsToStorageBucket adds the user-defined tags in the Infrastructure resource
+// to the passed GCP bucket resource. It's wrapper around addUserTagsToStorageBucket()
+// and additionally updates status condition in image registry Config resource.
+func (t *tagServiceManager) AddTagsToStorageBucket(ctx context.Context, cr *imageregistryv1.Config) error {
+	if err := t.addTagsToStorageBucket(ctx, cr); err != nil {
+		if errors.Is(err, NoTagsDefined) {
+			util.UpdateCondition(cr, defaults.StorageTagged, operatorapi.ConditionFalse,
+				gcpTagsSuccessStatusReason, NoTagsDefined.Error())
+			return nil
+		}
+		util.UpdateCondition(cr, defaults.StorageTagged, operatorapi.ConditionTrue,
```

Comment:
should the condition really be true here, i.e `operatorapi.ConditionTrue`? 
since there is an error, I'd expect the condition to be `operatorapi.ConditionFalse`. 
WDYT?

Reply:
I am using different reasons for success and failure which are `SuccessTaggingBucket` and `ErrorTaggingBucket`. And when there is an error, `ErrorTaggingBucket` reason is used and condition is set to true. Would it better if I just use `SuccessTaggingBucket` and set condition to `false` and message with `error string` when an error happens?

Reply:
Hmm I'm not sure I follow, maybe I'm misunderstanding things.
Doesn't `defaults.StorageTagged` condition set to `operatorapi.ConditionTrue` read as "the storage was tagged"? This is my confusion, please help me understand 😅

Reply:
Yeah, this will cause confusion. I was just reading reason and status and thought it to be apt and forgot about condition. Setting status to false during error scenario is the correct thing. I will update the code :)